### PR TITLE
feat: add navigatorContextProvider option,getter,extra

### PR DESCRIPTION
**Pull request by louis-bompart**
Add the concept of `navigatorContextProvider`.
The goal is to allow implementers to 'provide' navigation context instead of Atomic/Headless 'guessing' it from the browser.

This PR does implement the APIs on the Engine to do it. The following one will use those APIs in the actions & reducers.

https://coveord.atlassian.net/browse/KIT-3024

### CODE REVIEW

**Review by Spuffynism — APPROVED** [review]

**Review by fbeaudoincoveo — APPROVED** [review]

File: packages/headless/src/app/engine.ts [hunk 1]
```diff
@@ -248,13 +261,17 @@ function buildCoreEngine<
...
   }
   const logger = thunkExtraArguments.logger;
-  const thunkExtraArgumentsWithRelay: ExtraArgumentsWithRelay & ExtraArguments =
-    {
-      ...thunkExtraArguments,
-      get relay() {
-        return getRelayInstanceFromState(engine.state);
-      },
-    };
+  const navigatorContextProvider =
```
[the remark this answers is not in the document]
Comment (louis-bompart) [reply packages/headless/src/app/engine.ts:264]: Yes 🤡 I need to change that indeed and ensure we maintain the status quo.

[the remark this answers is not in the document]
Comment (louis-bompart) [reply packages/headless/src/app/engine.ts:264]: - [ ] ensure we don't break NodeJS

**Comment by louis-bompart** [comment]
Adding you @samisayegh as a reviewer, as we discussed that topic in length, it only seems natural ;) 

Importantly, we can't feed that information to `@coveo/relay` as it stands, and this should be tackled in subsequent PRs.

File: packages/headless/src/app/navigatorContextProvider.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+}
+
+export type NavigatorContextProvider = () => NavigatorContext;
+
+export const defaultBrowserNavigatorContextProvider: NavigatorContextProvider =
+  () => ({
+    referrer: document.referrer,
+    userAgent: navigator.userAgent,
+    location: window.location.href,
+  });
```
Comment (samisayegh) [inline packages/headless/src/app/navigatorContextProvider.ts:14]: An alternative implementation could be to use `relay.getMeta('')` here. The library is isomorphic, and will return `null` when used in a nodejs context. e.g.

```
const {referrer, userAgent, location} = relay.getMeta('')
```

**Review by samisayegh — APPROVED** [review]
Nice ✔️

**Pull request merged**